lib/Macros/Sources/SwiftMacros/DebugDescriptionMacro.swift

```diff
@@ -0,0 +1,312 @@
+        self.message = value
+    }
+    var message: String
+    var diagnosticID: MessageID { .init(domain: "DebugDescription", id: "DebugDescription")}
+    var severity: DiagnosticSeverity { .error }
+}
+
+extension MacroExpansionContext {
+    fileprivate func diagnose(node: some SyntaxProtocol, message: any DiagnosticMessage) {
+        diagnose(Diagnostic(node: node, message: message))
+    }
+}
+
+// MARK: - AST Helpers
+
+private func extractModuleName(from fileID: ExprSyntax) -> String? {
+    if let fileID = fileID.as(StringLiteralExprSyntax.self)?.representedLiteralValue,
+       let firstSlash = fileID.firstIndex(of: "/") {
+        return String(fileID.prefix(upTo: firstSlash))
+    }
+    return nil
+}
+
+private func identifier<Decl: DeclSyntaxProtocol>(of decl: Decl) -> TokenSyntax? {
+    decl.as(StructDeclSyntax.self)?.name ??
+    decl.as(ClassDeclSyntax.self)?.name ??
+    decl.as(EnumDeclSyntax.self)?.name ??
+    decl.as(ExtensionDeclSyntax.self)?.extendedType.as(IdentifierTypeSyntax.self)?.name
+}
+
+private func memberBlock<Decl: DeclSyntaxProtocol>(_ decl: Decl) -> MemberBlockSyntax? {
+    decl.as(StructDeclSyntax.self)?.memberBlock ??
+    decl.as(ClassDeclSyntax.self)?.memberBlock ??
+    decl.as(EnumDeclSyntax.self)?.memberBlock ??
+    decl.as(ExtensionDeclSyntax.self)?.memberBlock
+}
+
+private func syntaxKeyword<Decl: DeclSyntaxProtocol>(for decl: Decl) -> String {
+    if decl.is(StructDeclSyntax.self) { return "struct" }
+    if decl.is(ClassDeclSyntax.self) { return "class" }
+    if decl.is(EnumDeclSyntax.self) { return "enum" }
+    if decl.is(ExtensionDeclSyntax.self) { return "extension" }
+    assertionFailure("expected struct/class/enum/extension")
+    return "declaration"
+}
+
+private func genericParameters<Decl: DeclSyntaxProtocol>(_ decl: Decl) -> GenericParameterListSyntax? {
+    decl.as(StructDeclSyntax.self)?.genericParameterClause?.parameters ??
+    decl.as(ClassDeclSyntax.self)?.genericParameterClause?.parameters ??
+    decl.as(EnumDeclSyntax.self)?.genericParameterClause?.parameters
+}
+
+private func bindingName(_ binding: PatternBindingSyntax) -> String? {
+    binding.pattern.as(IdentifierPatternSyntax.self)?.identifier.text
+}
+
+private func isComputedProperty(_ binding: PatternBindingSyntax) -> Bool {
+    guard let accessors = binding.accessorBlock?.accessors else {
+        // No accessor block, not computed.
+        return false
+    }
+
+    switch accessors {
+    case .accessors(let accessors):
+        return accessors.contains { $0.accessorSpecifier.tokenKind == .keyword(.get) }
+    case .getter:
+        return true
+    @unknown default:
+        return false
+    }
+}
+
+// Enumerate a MemberAccess expression to produce DeclReferences in left to right order.
+// `a.b.c` will result in callbacks for each DeclReference `a`, `b`, and then `c`.
+private func enumerateMembers(_ memberAccess: MemberAccessExprSyntax, _ action: (DeclReferenceExprSyntax) -> Void) {
```

Comment:
I generally find code a lot easier to read if this would just return a `[DeclReferenceExprSyntax]` instead of taking a closure.

Reply:
good call.